noir-projects/aztec-nr/aztec/src/macros/dispatch/mod.nr

```diff
@@ -0,0 +1,141 @@
+use super::utils::compute_fn_selector;
+
+/// Returns an `fn public_dispatch(...)` function for the given module that's assumed to be an Aztec contract.
+pub comptime fn generate_public_dispatch(m: Module) -> Quoted {
+    let functions = m.functions();
+    let functions = functions.filter(|function: FunctionDefinition| function.has_named_attribute("public"));
+
+    let unit = get_type::<()>();
+
+    let ifs = functions.map(
+        |function: FunctionDefinition| {
+        let name = function.name();
+        let parameters = function.parameters();
+        let return_type = function.return_type();
+
+        let selector: Field = compute_fn_selector(function);
+
+        let mut parameters_size = 0;
+        for param in parameters {
+            parameters_size += size_in_fields(param.1);
+        }
+
+        let initial_read = if parameters.len() == 0 {
+            quote {}
+        } else {
+            // The initial calldata_copy offset is 1 to skip the Field selector
+            quote {
+                let input_calldata: [Field; $parameters_size] = dep::aztec::context::public_context::calldata_copy(1, $parameters_size);
+                let mut reader = dep::aztec::protocol_types::utils::reader::Reader::new(input_calldata);
+            }
+        };
+
+        let mut parameter_index = 0;
+        let reads = parameters.map(|param: (Quoted, Type)| {
+            let param_name = f"arg{parameter_index}".quoted_contents();
+            let param_type = param.1;
+            let read = quote {
+                let $param_name: $param_type = reader.read_struct(dep::aztec::protocol_types::traits::Deserialize::deserialize);
+            };
+            parameter_index += 1;
+            quote { $read }
+        });
+        let read = reads.join(quote { });
+
+        let mut args = &[];
+        for parameter_index in 0..parameters.len() {
+            let param_name = f"arg{parameter_index}".quoted_contents();
+            args = args.push_back(quote { $param_name });
+        }
+
+        let args = args.join(quote { , });
+        let call = quote { $name($args) };
+
+        let return_code = if return_type == unit {
+            quote { $call }
```

Comment:
If we don't do a return here, won't we continue evaluating the rest of the jump table given there's no `else` clauses?

Reply:
(also why no return?)

Reply:
I'm filling in for the Noir team here. This is basically their design of this function.

The dispatch function needs to return many different types depending on the function that it calls. Not even the serialized size is fixed. Noir cannot handle this, but the AVM can.

I created an opcode/oracle `avm_return` which does an EARLY return, and sets the right returndata.

That's also why the other ifs are not evaluated. BTW, the code was originally if-elses, but that broke Noir's parser.

Reply:
Discussed on Slack, there was a forced return missing for the void case, since we made all the if-elses ifs.